### ISSUE TITLE
serious performance issue with msh.mesh_io.read_msh

### ISSUE THREAD

**Issue by octomike** [simnibs/simnibs#33]
Hi, I am currently hunting down a performance issue following a call to `headreco all sub-foo sub-foo_T1w.nii`.

With my test subject foo the current simnibs 3.1.2 / gmsh 3.0.6 produce a mesh file containing 864030 vertices 5908334 elements in binary msh v2 format. However the `$Elements` data appear to individually grouped in the file, resulting in around 6 million calls to [np.fromfile](https://github.com/simnibs/simnibs/blob/master/simnibs/msh/mesh_io.py#L4113), reading 3 integers a time. While this might work okay in a local/SSD scenario it's absolutely prohibitive on our network file system and we had to cancel our jobs after 24 hours of reading the meshes into memory. Here are some benchmarks for a simple read using the file:

```python
from simnibs import mesh_io
import time

start = time.time()
m = mesh_io.read_msh('sub-foo.msh')
stop = time.time()
print(stop - start)
```

Storage Type | Measured Times (seconds)
 -- | -- 
/dev/shm | 254
SAS-Raid1 | 251
local SSD | 320 (slow CPU)
NFS-10Gbit | 1662, 2127,1689
Beegfs (native cache, experimental and broken) | 2477, 2052
Beegfs (default, our scenario unfortunately) | 86400+ (interrupted)

To make things more confusing during my tests I also ended up with a binary msh file that seemed to have the 6 million elements grouped into two groups. Hence only two `np.fromfile()` calls were necessary and the file was read in **under a second**.

Initially I came up with a minimally invasive workaround to swap `np.fromfile` with `np.frombuffer` wasting a little memory and reading the file into memory right in the beginning. My two questions now are:

1. Can you shed some light on why gmsh sometimes produces a perfectly manageable file?
2. Would you consider a PR with the in-memory IO workaround?

**Comment by guilhermebs** [comment]
I agree that this function has a performance problem in _some_ .msh files, which are structured with several $Elements sections. Unfortunately this is how Gmsh saves files after meshing, (I really don't know why, it looks very inefficient). However, once the file is read by SimNIBS in and saved again, it contains only 2 $Elements sections (one for all triangles and another for all tetrahedron), so I/O becomes very fast. 

1. The reason why you get fast I/O in some files is because those were saved by SimNIBS, not Gmsh. See the explanation above
2. In the near future we will no longer use Gmsh for meshing, so we won't have to read in these inefficient msh files. My suggestion is that you try to incorporate a step in your pipeline where you read in the .msh file and save it again, using either matlab or python. But you can send the PR as well.

**Comment by octomike** [comment]
Hi, thanks for your feedback! I'm still looking into the lowest impact solution to this problem, before pushing some python hacks..

I wondered if maybe gmsh can be conviced to group the elements, but apparently not for the msh2 format:

  https://gitlab.onelab.info/gmsh/gmsh/-/issues/964#note_10923

Since I saw there is a `_read_msh_4` function in msh_io, I tried saving to 4.0 as the gmsh developer suggested with: `gmsh -format msh40 -save -n -o new.msh old.msh` and indeed the I/O pattern for reading this file is much better:

```
% time     seconds  usecs/call     calls    errors syscall                                                                     
------ ----------- ----------- --------- --------- ----------------                                                            
 72.50    2.465572       42510        58           futex                                                                       
 15.36    0.522465          23     22724           sched_yield                                                                 
  4.11    0.139597          69      2033           read         <--- yeah
[...]
```
Single line msh 2.0 reading for comparison (interrupted..)

```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 87.21  291.914399         291   1003769           read   <--- ouch
  4.90   16.407812           6   2978020         8 lseek
  4.56   15.250845          30    502514           fstat
```


So I guess saving the mesh in 4.0 format in the first place seems to be the obvious workaround. **However** the bundled gmsh is version 3.0.6 (November 2017) and it looks like it's the last version that is *not* supporting the 4.0 msh format. :sob:  Do you think it would be possible to bundle a (any) 4.x version of gmsh? Then you could save in 4.0, eliminating the performance issue at the core.

**Comment by vfonov** [comment]
one can also copy .msh file to the node-local storage before running simnibs

**Comment by octomike** [comment]
> one can also copy .msh file to the node-local storage before running simnibs

True. Right now we are doing everything in `/dev/shm` by moving the nifti file there - But I would like to avoid such hacks in the headreco pipeline if possible.

**Comment by vfonov** [comment]
by the way, I implemented  a c++ version of .msh reader/writer for libigl : https://github.com/libigl/libigl/pull/1578 , exactly for the purpose of dealing with files generated by simnibs/gmsh 3.0  - maybe that reader would work faster on the network storage.

**Comment by guilhermebs** [comment]
@octomike we have not changed to Gmsh 4 because we could not test the meshing well enough. Meshing is a quite delicate procedure, so we really need to test it a lot to make sure it is stable. By the next release we will likely not be using Gmsh for meshing anymore, just for visualization. You could try to change the Gmsh version in your local install and see if it works. Just move the gmsh binary to `simnibs/bin/linux/`

**Comment by Kstoffer** [comment]
After discussing this issue with Axel and Oula at our meeting yesterday, and testing on our local network cluster we decided to implement a buffered read flag to the read_msh function.  Quick testing suggests that on our network it is around 15 times faster or so for the problematic files. The flag (buffered) will default to False as for cleaned up meshes it is not faster and will use more memory. We also changed the calls that use the inefficient gmsh v2 output such that it should use buffering in the case where it is relevant.
More speedup is definitely possible but the multiple read requests over network should at least be eliminated. Anecdotally, we actually had a more efficient implementation of this earlier which would try to predict the number of elements and back track if necessary, but the code was deprecated due to bad readability, the hack-ish feeling of it, and the fact that it is not really an issue with cleaned up files. JIT’ing this part of the code is also an option but we are currently not ready to add such a dependency right now.
The build actually failed on MacOS X 10.14 but I don't think that has anything to do with the edit, we will investigate this further, also we will do a more comprehensive test of the full headreco pipeline to ensure it did not break anything before finally making this part of a release.
